### ISSUE TITLE
Deploying to Heroku

### ISSUE THREAD

**Issue by deg** [Day8/re-frame-template#32]
I've just written a toy app using this template, and I'd like to deploy it to Heroku.
I followed [the usual steps](https://devcenter.heroku.com/articles/getting-started-with-clojure#deploy-the-app) but hit a failure:

````
...
2016-08-02T09:41:59.777243+00:00 heroku[slug-compiler]: Slug compilation started
2016-08-02T09:41:59.777249+00:00 heroku[slug-compiler]: Slug compilation finished
2016-08-02T09:42:04.537917+00:00 heroku[web.1]: Starting process with command `lein with-profile production trampoline run`
2016-08-02T09:42:06.800054+00:00 app[web.1]: Setting JAVA_TOOL_OPTIONS defaults based on dyno size. Custom settings will override them.
2016-08-02T09:42:06.813188+00:00 app[web.1]: Picked up JAVA_TOOL_OPTIONS: -Xmx350m -Xss512k -Dfile.encoding=UTF-8
2016-08-02T09:42:45.664277+00:00 app[web.1]: Compiling ClojureScript...
2016-08-02T09:42:46.963767+00:00 app[web.1]: Picked up JAVA_TOOL_OPTIONS: -Xmx350m -Xss512k -Dfile.encoding=UTF-8
2016-08-02T09:42:48.681490+00:00 app[web.1]: 	at clojure.lang.Compiler.load(Compiler.java:7391)
2016-08-02T09:42:48.681325+00:00 app[web.1]: Exception in thread "main" java.io.FileNotFoundException: Could not locate cljsbuild/compiler__init.class or cljsbuild/compiler.clj on classpath., compiling:(/tmp/form-init9197767631145422317.clj:1:73)
2016-08-02T09:42:48.681536+00:00 app[web.1]: 	at clojure.lang.Compiler.loadFile(Compiler.java:7317)
2016-08-02T09:42:48.681592+00:00 app[web.1]: 	at clojure.main$load_script.invokeStatic(main.clj:275)
...
````

I see in this project's [README.md](https://github.com/Day8/re-frame-template/blob/master/README.md) the note `Desired profiles: +Heroku`, so I assume it is known that some magic is needed to make this work. But, I don't know what the next step is, or even exactly what the error means.  Help?  Thanks!

**Comment by deg** [comment]
FWIW, I verified that I can run the uberjar locally, or in a docker instance, with a simple
````
lein uberjar
java -jar flashcards-0.1.0-SNAPSHOT-standalone.jar
````

**Comment by gadfly361** [comment]
@deg Unfortunately, I am unfamiliar with using heroku. Regarding the note in the README, I figured someone in the community would submit a PR to add the `+heroku` profile.  

I know [reagent-template](https://github.com/reagent-project/reagent-template) supports heroku, perhaps what is needed can be cleaned from there.

**Comment by deg** [comment]
Doing a quick copy from reagent-template, I got a bit closer.

The problem seems to be that I was missing a `Procfile`. So, cribbing shamelessly, I created:
````
web: java $JVM_OPTS -cp target/flashcards-0.1.0-SNAPSHOT-standalone.jar clojure.main -m flashcards.server
````

This let me start locally with `foreman start`, with only one surprising quirk - the default port moved from 3000 to 5000.

So, feeling hopeful, I pushed to Heroku, but

````
...
2016-08-02T20:06:53.246251+00:00 heroku[web.1]: Starting process with command `java $JVM_OPTS -cp target/flashcards-0.1.0-SNAPSHOT-standalone.jar clojure.main -m flashcards.server`
2016-08-02T20:06:55.762757+00:00 app[web.1]: Setting JAVA_TOOL_OPTIONS defaults based on dyno size. Custom settings will override them.
2016-08-02T20:06:55.780069+00:00 app[web.1]: Picked up JAVA_TOOL_OPTIONS: -Xmx350m -Xss512k -Dfile.encoding=UTF-8
2016-08-02T20:06:56.214128+00:00 app[web.1]: Error: Could not find or load main class clojure.main
2016-08-02T20:06:56.320162+00:00 heroku[web.1]: Process exited with status 1
2016-08-02T20:06:56.313475+00:00 heroku[web.1]: State changed from starting to crashed
...
````

So, looks like something in my project replaces `clojure.main`

I had built with 
````
lein new re-frame flashcards +cider +test +compojure +garden +re-com +secretary
````
and my guess would be that this is compojure-related. But, I'm not familiar enough, and it's too late at night here, so I'm going to give this a rest for now.

_(I'll be a very happy camper in the morning if someone recognizes the issue and surprises me with an answer.  Pretty please?)_

**Comment by danielcompton** [comment]
There's three things you need to do:

1. Create your template with `+handler`, not `+compojure`
2. Add a line in `project.clj`: `:uberjar-name "flashcards.jar"`
3. Make sure your `Procfile` says `web: java $JVM_OPTS -cp target/flashcards.jar clojure.main -m flashcards.server`

This worked for me, and I was able to deploy to Heroku and run it.

**Comment by deg** [comment]
tl;dr - Thanks!  I've got it almost working now; everything but the css.

Details:
- "+handler" vs "+compojure". Red herring. I had fixed this earlier, but had cut-n-pasted from my earlier notes. I've edited my comment above now, in a perhaps futile attempt to minimize confusion of future readers.
- :uberjar-name - Added now, and this seems to have fixed the problem. But, I don't understand why this worked. As far as I can see, my previous Procfile was essentially the same, and just referred to the default name of the uberjar file.  Is the contents of the jar file somehow different when uberjar-name is specified explicitly?
- But, the heroku site is not getting my css. It is being seen by locally, including by `foreman start`. I will look into this right now, but my guess is that build is not including it, but the foreman run is picking up a local copy of the css that I had compiled earlier. (I'm surprised that it was not removed by `lein clean`).  I'll post back when/if I find the problem, unless someone beats me to the punch.

**Comment by deg** [comment]
Re the css problem: I'm confused.
If I do a `lein clean && lein garden once`, the css is generated.  But then, if I do `lein uberjar`, the css file is deleted.

I don't understand enough of what is going in in `project.clj` to debug this competently.

**Comment by danielcompton** [comment]
You need to add your css compile to the prep tasks.

**Comment by deg** [comment]
Is this working for you? Can you send me the exact line you use?

When I try, with variations of 
````
  :prep-tasks [["cljsbuild" "once" "min"] ["garden" "once"] "compile"]
````
I get an endless loop of `Compiling Garden...`, like what was reported at [https://github.com/noprompt/lein-garden/issues/43](https://github.com/noprompt/lein-garden/issues/43) or an equally endless loop of `Compiling Garden...  Compiling ClojureScript...`

**Comment by danielcompton** [comment]
> :uberjar-name - Added now, and this seems to have fixed the problem. But, I don't understand why this worked. As far as I can see, my previous Procfile was essentially the same, and just referred to the default name of the uberjar file. Is the contents of the jar file somehow different when uberjar-name is specified explicitly?

When you create an uberjar, by default two files are created. `app-version.jar`, and `app-version-standalone.jar`. You need to use the standalone one as this is the uberjar with all the classes in it. Setting the uberjar name just sets it to a standard name so you don't need to update your Procfile when your app version changes.

> Is this working for you? Can you send me the exact line you use?

I didn't do anything with Garden when I deployed to Heroku. I've never used it, so I can't help you there sorry.

The issues you're grappling with are a bit broader than just related to re-frame template. I recommend you ask for help on the Clojure mailing list or Clojurians Slack, as there are people there who may be able to help you further. Thanks!

**Comment by gadfly361** [comment]
@deg I just tried bumping the lein-garden version to `0.2.8` and that did the trick!  App is deployed to heroku with css from garden.

**Comment by deg** [comment]
Thanks!  I'm good now.

I've not tested enough to create a PR but, for the sake of the next person reading this, here is a summary of the changes I had to make.

This was for a project created with
````
lein new re-frame myproject +cider +test +handler +garden +re-com +routes
````

I had to:

- Add a file `Procfile`, containing
````
web: java $JVM_OPTS -cp target/myproject.jar clojure.main -m myproject.server
````

- Add to `project.clj`:
````
  :uberjar-name "myproject.jar"
````

- Change two lines in `project.clj`, for `garden` to run as part of the build, and use a working version:
````
  :plugins [[lein-cljsbuild "1.1.3"]
            [lein-garden "0.2.8"]]
  ...
  :prep-tasks [["cljsbuild" "once" "min"] ["garden" "once"] "compile"]
````

@gadfly361 and @danielcompton - thank you agan for all your help!

**Comment by gadfly361** [comment]
Hmm I was able to push to heroku once with garden's css. However, since I have been getting an error like this after I run `git push heroku master`:

**Comment by gadfly361** [comment]
Looks like this can be resolved by manually including `[ns-tracker "0.3.0"]` in the :dependencies vector in project.clj. I was trying to create a minimal example, and I am guessing that one of the extra +xxx profiles was bringing in ns-tracker 0.3.0.

**Comment by gadfly361** [comment]
`+handler` profile should now support heroku and take into account the +less and +garden profiles

**Comment by manandearth** [comment]
@deg 
as for:
>   :prep-tasks [["cljsbuild" "once" "min"] ["garden" "once"] "compile"]

It seems that with  `[garden " 1.3.6"]` and plugins: `[[lein-cljsbuild "1.1.7"]  [lein-garden "0.3.0"]]`  
garden is being compiled with the cljsbuild and the `"garden" "once"` only empties the css file again... 
I was puzzled over this but found that it deploys fine without the garden command:

Hence:
 `:prep-tasks [["cljsbuild" "once" "min"] "compile"]
